File: labs/lingo/task3/turing.py
from typing import Dict, List, Literal, TypedDict

Direction = Literal["L", "R", "N"]

class TuringRule(TypedDict):
    write: str
    move: Direction
    next: str

TuringProgram = Dict[str, TuringRule]

class TuringMachine:
    FINAL_STATE = "halt"

    def __init__(self, initial_tape: str, program: TuringProgram):
        self.tape: List[str] = list(initial_tape)
        self.head: int = 0
        self.state: str = "q0"
        self.program: TuringProgram = program
        self.history: List[str] =[]

    @property
    def current_symbol(self) -> str:
        if 0 <= self.head < len(self.tape):
            return self.tape[self.head]
        return "_"

    def step(self) -> bool:
        if self.state == self.FINAL_STATE:
            return False
            
        key = f"{self.state}:{self.current_symbol}"
        rule = self.program.get(key)

        if not rule:
            self.state = self.FINAL_STATE
            return False

File: labs/lingo/task3/turing.ts
export type Direction = "L" | "R" | "N";

export interface TuringRule {
  write: string;
  move: Direction;
  next: string;
}

export type TuringProgram = Record<string, TuringRule>;

export class TuringMachine {
  tape: string[];
  head: number;
  state: string;
  program: TuringProgram;
  history: string[] = [];

  private readonly FINAL_STATE = "halt";

  constructor(initialTape: string, program: TuringProgram) {
    this.tape = initialTape.split("");
    this.head = 0;
    this.state = "q0";
    this.program = program;
  }

  get currentSymbol() {
    return this.tape[this.head] || "_";
  }

  step(): boolean {
    if (this.state == this.FINAL_STATE) return false;
    const key = `${this.state}:${this.currentSymbol}`;
    const rule = this.program[key];

    if (!rule) {
      this.state = this.FINAL_STATE;
      return false;
    }

    this.tape[this.head] = rule.write;
    if (rule.move == "L") {
      this.head--;
      if (this.head < 0) {
        this.tape.unshift("_");
        this.head = 0;
      }
    } else if (rule.move == "R") {
      this.head++;
      if (this.head >= this.tape.length) {
        this.tape.push("_");
      }
    }
    this.state = rule.next;

    return this.state != this.FINAL_STATE;
  }

  getTapeString() {
    return this.tape.join("");
  }
}
